executable-monitor/executable-monitor.py

```diff
@@ -0,0 +1,140 @@
+#!/usr/bin/env python3
+
+import os, sys
+from argparse import ArgumentParser
+import subprocess
+import time
+
+if __name__ == '__main__':
+    parser = ArgumentParser(description='Executable monitor.')
+    parser.add_argument('--exe-path',
+                        type=str,
+                        required=True,
+                        help='Path to the executable.')
+    parser.add_argument('--log-dir',
+                        type=str,
+                        required=True,
+                        help='Path to directory to store logs in.')
+    parser.add_argument('--timeout-seconds',
+                        type=int,
+                        required=True,
+                        help='Timeout for each executable run.')
+    parser.add_argument('--number-of-runs',
+                        type=int,
+                        required=True,
+                        help='Timeout for each executable run.')
+    parser.add_argument('--success-line',
+                        type=str,
+                        required=False,
+                        help='Line that indicates executable completed successfully. Required if --success-exit-status is not used.')
+    parser.add_argument('--success-exit-status',
+                        type=int,
+                        required=False,
+                        help='Exit status that indicates that the executable completed successfully. Required if --success-line is not used.')
+
+    args = parser.parse_args()
+
+    if args.success_exit_status is None and args.success_line is None:
+        print("Must specify at least one of the following: --success-line, --success-exit-status.")
+        sys.exit(1)
+
+    if not os.path.exists(args.exe_path):
+        print(f'Input executable path \"{args.exe_path}\" does not exist.')
+        sys.exit(1)
+
+    # Create log directory if it does not exist.
+    if not os.path.exists(args.log_dir):
+        os.makedirs(args.log_dir, exist_ok = True)
+
+    # Convert any relative path (like './') in passed argument to absolute path.
+    EXE_PATH = os.path.abspath(args.exe_path)
+    LOG_DIR = os.path.abspath(args.log_dir)
+
+    print(f"Running executable: {EXE_PATH} ")
+    print(f"Storing logs in: {LOG_DIR}")
+    print(f"Timeout per run (seconds): {args.timeout_seconds}")
+    print(f"Number of runs: {args.number_of_runs}\n")
+
+    EXE_NAME = os.path.basename(EXE_PATH)
+
+    log_file = open(f'{LOG_DIR}/{EXE_NAME}_output.txt', 'w')
+
+    exe = subprocess.Popen([EXE_PATH], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True)
```

Comment:
Should there be a check here that the executable actually launches before moving forward? Thinking about a flow of execution where for some reason this doesn't launch correctly and we wait 5 minutes in a github workflow for each of these demos to silently fail

Reply:
The exe.poll call will get the return value from the executable and the loop below exits if the executable has returned. If one calls subprocess.popen with a file that isn't an executable, it throws an exception and the program stops.

Reply:
Additionally, it is checked above that the executable path exists.